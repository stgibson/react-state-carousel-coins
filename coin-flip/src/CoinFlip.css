.CoinFlip {
  display: flex;
  flex-direction: column;
}

.CoinFlip-button {
  width: 90px;
  /* Learned how to center button with fixed width at https://stackoverflow.com/questions/19764284/how-to-center-a-fixed-width-button-horizontally */
  margin: 0 auto;
}