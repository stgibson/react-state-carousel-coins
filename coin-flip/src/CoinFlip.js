import React, { useState } from "react";
import "./CoinFlip.css";
import Coin from "./Coin";

const CoinFlip = () => {
  const [side, setSide] = useState(null);
  const [numFlips, setNumFlips] = useState(0);
  const [numHeads, setNumHeads] = useState(0);
  const [numTails, setNumTails] = useState(0);

  // flips coin by random
  const flipCoin = () => {
    const flip = Math.random() < 0.5;
    setNumFlips(numFlips + 1);
    if (flip) {
      setSide("h");
      setNumHeads(numHeads + 1);
    }
    else {
      setSide("t");
      setNumTails(numTails + 1);
    }
  };

  return (
    <div className="CoinFlip">
      <h1>Let's flip a coin!</h1>
      { side && <Coin side={ side } /> }
      <button className="CoinFlip-button" onClick={ flipCoin }>FLIP MEEE</button>
      <p>Out of { numFlips } flips, there have been { numHeads } heads 
      and { numTails } tails.</p>
    </div>
  );
};

export default CoinFlip;