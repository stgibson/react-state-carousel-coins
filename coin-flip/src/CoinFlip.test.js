import React from "react";
import { render, fireEvent } from "@testing-library/react";
import CoinFlip from "./CoinFlip";

beforeEach(function() {
  jest
    .spyOn(Math, "random")
    .mockReturnValueOnce(0.25)
    .mockReturnValueOnce(0.75);
});

it("loads without crashing", function() {
  render(<CoinFlip />);
});

it("matches snapshot as heads", function() {
  const { asFragment } = render(<CoinFlip />);
  expect(asFragment()).toMatchSnapshot();
});

it("doesn't show coin when the page loads", function() {
  const { queryByText } = render(<CoinFlip />);
  
  // verify both sides of coin not on page
  expect(queryByText("Heads")).not.toBeInTheDocument();
  expect(queryByText("Tails")).not.toBeInTheDocument();
});

it("updates text correctly on heads", function() {
  const { queryByText } = render(<CoinFlip />);

  expect(queryByText("Out of 0 flips, there have been 0 heads and 0 tails."))
    .toBeInTheDocument();

  // click button to get heads
  const button = queryByText("FLIP MEEE");
  fireEvent.click(button);
  expect(queryByText("Out of 1 flips, there have been 1 heads and 0 tails."))
    .toBeInTheDocument();
});

it("updates text correctly on tails", function() {
  const { queryByText } = render(<CoinFlip />);

  expect(queryByText("Out of 0 flips, there have been 0 heads and 0 tails."))
    .toBeInTheDocument();

  // click button twice to get tails
  const button = queryByText("FLIP MEEE");
  fireEvent.click(button);
  fireEvent.click(button);
  expect(queryByText("Out of 2 flips, there have been 1 heads and 1 tails."))
    .toBeInTheDocument();
});

afterEach(function() {
  Math.random.mockRestore();
});