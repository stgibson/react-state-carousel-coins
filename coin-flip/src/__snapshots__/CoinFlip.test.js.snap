// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot as heads 1`] = `
<DocumentFragment>
  <div
    class="CoinFlip"
  >
    <h1>
      Let's flip a coin!
    </h1>
    <button>
      FLIP MEEE
    </button>
    <p>
      Out of 0 flips, there have been 0 heads and 0 tails.
    </p>
  </div>
</DocumentFragment>
`;
